webapp/src/Form/Type/UserRegistrationType.php

```diff
@@ -85,6 +87,39 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'mapped' => false,
             ]);
 
+        $registrationCategoryNames = $this->dj->dbconfig_get('registration_category_names', []);
+        $showCategoryChoice        = count($registrationCategoryNames) !== 1;
+        $registrationCategories    = [];
+        foreach ($registrationCategoryNames as $name) {
+            $category = $this->em->getRepository(TeamCategory::class)->findOneBy(['name' => $name]);
+            // silently ignore invalid category names
+            if ($category) {
+                $registrationCategories[] = $category;
+            }
+        }
+        // if all the category names are invalid, the user won't be able to fill in the form; don't even show it
+        if (empty($registrationCategories)) {
+            throw new HttpException(500, 'Self-registration team categories are invalid');
+        }
+
+        $builder
```

Comment:
I'd say it is indeed better to not add this element to the form if `$registrationCategories` contains only one item and then in the `SecurityController` do a check for if it is only one item and, if so, use that.

Reply:
I've changed it to conditionally add the form element, but I think it should still be based on the number of categories in the configuration option (`$registrationCategoryNames`) rather than how many were found in the database (`$registrationCategories`).
Suppose the admin sets the option to two names and one of them is invalid. The intention is clearly to give users a choice. A dropdown with one category seems silly, but it will give a strong hint that the other category is invalid. Removing the dropdown entirely would be very confusing.
There is only a difference when the configuration option is invalid, and I think in that situation debuggablity is more important than presentation. Does that sound reasonable?

[Full diff](https://github.com/tom93/domjudge/compare/self-registration-multiple-categories-v1-hide-with-html-attribute..tom93:self-registration-multiple-categories-v2-conditionally-build-field), prettified diff:

```diff
--- a/webapp/src/Controller/SecurityController.php
+++ b/webapp/src/Controller/SecurityController.php
@@ -125,2 +125,11 @@ class SecurityController extends AbstractController
             $teamName = $registration_form->get('teamName')->getData();
+
+            if (count($registrationCategoryNames) === 1) {
+                $teamCategory = $em->getRepository(TeamCategory::class)->findOneBy(['name' => $registrationCategoryNames[0]]);
+                if (!$teamCategory) {
+                    throw new HttpException(500, sprintf("Team category '%s' not found", $registrationCategoryNames[0]));
+                }
+            } else {
+                // count($registrationCategoryNames) > 1, 'teamCategory' field exists
                 $teamCategory = $registration_form->get('teamCategory')->getData();
+            }
--- a/webapp/src/Form/Type/UserRegistrationType.php
+++ b/webapp/src/Form/Type/UserRegistrationType.php
@@ -88,4 +88,4 @@ class UserRegistrationType extends AbstractType
         $registrationCategoryNames = $this->dj->dbconfig_get('registration_category_names', []);
-        $showCategoryChoice        = count($registrationCategoryNames) !== 1;
+        if (count($registrationCategoryNames) > 1) {
             $registrationCategories = [];
             foreach ($registrationCategoryNames as $name) {
@@ -108,12 +108,11 @@ builder->add('teamCategory', EntityType::class, [
                     'choice_label' => 'name',
-                    'placeholder' => $showCategoryChoice ? '-- Select category --' : false,
+                    'placeholder' => '-- Select category --',
                     'choices' => $registrationCategories,
                     'attr' => [
                         'placeholder' => 'Category',
-                    ] + (
-                        $showCategoryChoice ? [] : ['hidden' => 'hidden']
-                    ),
+                    ],
                     'constraints' => [
                         new NotBlank(),
                     ],
                 ]);
+        }
```

Reply:
I think that if you want any errors to be debuggable, then it's better to write a notice level log message and/or add a configuration check to see that all categories exist. But I think a team user should not be presented with a weird selection dialog with only one (or even zero) options.

Reply:
Updated to make all the decisions based on the number of categories that exist in the database.
[Diff from previous version](https://github.com/DOMjudge/domjudge/compare/59e33099b6e08a745b743b319a9a98dc5100898e..8e46632f2e7f6a6c70dab688ec3578f6a0e19742)
I already added a configuration check to see that all the categories exist.
The previous version avoided showing a selection dialog with zero options by explicitly throwing an exception.